playing-card-detector/python/offlineCvDev.py

```diff
@@ -31,76 +60,53 @@ def main():
 # suit is S for spades, C for clubs, D for diamonds and H for hearts,
 # rank is a number 1-13 where 1 is an ace and 13 is a king.
 def analyzeImage(image):
-    subImages = splitImage(image)
+    subImages = []
+    subImages.append(image[:, 0:int(RECT_WIDTH)])
+    subImages.append(image[:, int(WIDTH-RECT_WIDTH) : WIDTH])
 
     resultString = ""
     for i, subImage in enumerate(subImages):
-        player = 1
-        if i > 5: player = 2
-        position = (i % 6) + 1
-        resultString += analyzeSubImage(subImage, player, position)
-
-    #cv2.imshow("image", image)
-    cv2.waitKey(0)
+        player = i + 1
+        resultString += analyzeSubImage(subImage, player)
 
     return resultString
 
-# Splits a full-size image into 12 separate sub-images, one for each position.
-def splitImage(image):
-    subImages = []
-
-    for i in range(12):
-        xOffset = 0
-        if i > 5: xOffset = WIDTH - RECT_WIDTH
-        yOffset = (i % 6) * RECT_HEIGHT
-        x1 = xOffset
-        x2 = xOffset + RECT_WIDTH
-        y1 = yOffset
-        y2 = yOffset + RECT_HEIGHT
-        subImages.append(image[y1:y2, x1:x2])
-
-    return subImages
-
 
-# Analyzes a single sub-image for the given player and position and returns the string for that particular sub-image.
-# If nothing is found, an empty string is returned.
-def analyzeSubImage(subImage, player, position):
+# Analyzes one player's part of the image with up to "CARD_AREAS" number of cards.
+# for each card, a string is created containing player_id, position, suit, rank, angle.
+# If no card is found, an empty string is returned.
+def analyzeSubImage(subImage, player):
     _, thresh = cv2.threshold(subImage,100,255,cv2.THRESH_BINARY)
     contours, hier = cv2.findContours(thresh,cv2.RETR_TREE,cv2.CHAIN_APPROX_SIMPLE)
     index_sort = sorted(range(len(contours)), key=lambda i : cv2.contourArea(contours[i]),reverse=True)
 
     # If there are no contours or no card is found, do nothing.
-    if len(contours) == 0 or not (CARD_AREA_MIN <= cv2.contourArea(contours[0]) <= CARD_AREA_MAX):
+    if len(contours) == 0:
         return ""
 
-    # Count contours that are large enough but not too small to be a suit marker.
-    rank = 0
-    for contour in contours:
-        if SUIT_AREA_MIN <= cv2.contourArea(contour) <= SUIT_AREA_MAX: rank += 1
+    resultString = ""
+    foundCards = 0
+    currentCardIndex = 0
+    while(foundCards < CARD_AREAS and currentCardIndex < len(contours) and isCardContour(contours[currentCardIndex])):
+        _, y, w, h = cv2.boundingRect(contours[currentCardIndex])
+        rotated = int(w < h)
+        position = int((y + h / 2) / RECT_HEIGHT)
+        rank = 0
+        suitMarkerIndex = currentCardIndex + 1
+        while(suitMarkerIndex < len(contours) and isSuitContour(contours[suitMarkerIndex])): 
```

Comment:
Does it make sense to traverse the contour hierarchy instead? Maybe this is robust enough since the contours are sorted?